src/gui/dialogs/units_dialog.hpp

```diff
@@ -0,0 +1,241 @@
+/*
+	Copyright (C) 2016 - 2024
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include "gui/dialogs/modal_dialog.hpp"
+#include "gui/widgets/button.hpp"
+#include "gui/widgets/group.hpp"
+#include "gui/widgets/listbox.hpp"
+#include "gui/widgets/toggle_button.hpp"
+#include "gui/widgets/unit_preview_pane.hpp"
+#include "team.hpp"
+#include "units/ptr.hpp"
+#include "units/types.hpp"
+#include "utils/optional_fwd.hpp"
+
+#include <boost/dynamic_bitset/dynamic_bitset.hpp>
+#include <optional>
+#include <string>
+#include <vector>
+
+class team;
+
+namespace gui2::dialogs
+{
+
+class units_dialog : public modal_dialog
+{
+public:
+	units_dialog();
+
+	int get_selected_index() const
+	{
+		return selected_index_;
+	}
+
+	bool is_selected() const
+	{
+		return selected_index_ != -1;
+	}
+
+	utils::optional<unit_const_ptr> get_unit() const
+	{
+		if(is_selected() || !recall_list_.empty()) {
+			return utils::optional{recall_list_[get_selected_index()]};
```

Comment:
You don't need to wrap these in `utils::optional{}`. Just do `return recall_list_[get_selected_index()];`. Sorry for not catching this before. The only time the return types need to be the same is if you're using a ternary operator.